junit-platform-commons/src/main/java/org/junit/platform/commons/support/conversion/TypeDescriptor.java

```diff
@@ -0,0 +1,113 @@
+/*
+ * Copyright 2015-2025 the original author or authors.
+ *
+ * All rights reserved. This program and the accompanying materials are
+ * made available under the terms of the Eclipse Public License v2.0 which
+ * accompanies this distribution and is available at
+ *
+ * https://www.eclipse.org/legal/epl-v20.html
+ */
+
+package org.junit.platform.commons.support.conversion;
+
+import static org.apiguardian.api.API.Status.EXPERIMENTAL;
+
+import java.lang.reflect.Field;
+import java.lang.reflect.Parameter;
+import java.util.NoSuchElementException;
+import java.util.Objects;
+import java.util.Optional;
+
+import org.apiguardian.api.API;
+import org.jspecify.annotations.Nullable;
+import org.junit.platform.commons.util.Preconditions;
+import org.junit.platform.commons.util.ReflectionUtils;
+
+/**
+ *
+ *
+ * @since 6.0
+ */
+@API(status = EXPERIMENTAL, since = "6.0")
+public final class TypeDescriptor {
+
+	/**
+	 * {@code TypeDescriptor} returned when no value is available.
+	 */
+	public static final TypeDescriptor NONE = new TypeDescriptor(null);
+
+	private final @Nullable Class<?> type;
+
+	public static TypeDescriptor forClass(Class<?> clazz) {
+		Preconditions.notNull(clazz, () -> "clazz must not be null");
+		return new TypeDescriptor(clazz);
+	}
+
+	public static TypeDescriptor forInstance(@Nullable Object instance) {
+		return instance != null ? forClass(instance.getClass()) : NONE;
+	}
+
+	public static TypeDescriptor forField(Field field) {
+		Preconditions.notNull(field, "field must not be null");
+		return forClass(field.getType());
+	}
+
+	public static TypeDescriptor forParameter(Parameter parameter) {
+		Preconditions.notNull(parameter, "parameter must not be null");
+		return forClass(parameter.getType());
+	}
+
+	private TypeDescriptor(@Nullable Class<?> type) {
+		this.type = type;
+	}
+
+	public Class<?> getType() {
+		if (type == null) {
+			throw new NoSuchElementException("No type present");
+		}
+		return type;
+	}
+
+	/**
+	 * Get the wrapper type of this type descriptor, if available.
+	 *
+	 * <p>If this type descriptor represents a primitive type, this method
+	 * returns the corresponding wrapped type. Otherwise, this method returns
+	 * {@link Optional#empty() empty()}.
+	 *
+	 * @return an {@code Optional} containing the wrapper type; never
+	 * {@code null} but potentially empty
+	 */
+	// FIXME [NullAway] parameter type of referenced method is @NonNull, but parameter in functional interface method java.util.function.Function.apply(T) is @Nullable
+	@SuppressWarnings("NullAway")
+	public Optional<Class<?>> getWrapperType() {
+		return Optional.ofNullable(type).map(ReflectionUtils::getWrapperType);
+	}
```

Comment:
I think this is a NullAway bug. I'll compose a small reproducer and raise an issue with the project.

Reply:
Hmm, that is weird. How does `NullAway` know about the nullability of the `Function` type parameters of `Optional.map`? Maybe it doesn't?

Reply:
It looks to me like it thinks `Optional.map` is typed as 

`<U> Optional<U> map(Function<? super @Nullable T, ? extends @Nullable U> mapper)`

when it should be

`<U> Optional<U> map(Function<? super @NonNull T, ? extends @Nullable U> mapper)`